File: backend/app/schemas/v1/course.py
from __future__ import annotations
from datetime import datetime
import uuid
from typing import Literal, Optional
from pydantic import BaseModel, Field


class CourseCreate(BaseModel):
    """POST /api/v1/courses — create a new course."""

    course_name: str = Field(min_length=1, max_length=255)
    course_code: Optional[str] = Field(None, max_length=50)
    department_id: Optional[uuid.UUID] = None
    room_id: Optional[uuid.UUID] = None
    teacher_id: Optional[int] = Field(None, description="Teacher ID from teachers table to assign as instructor")
    attendance_mode: Literal["preset", "flexible", "custom"] = "preset"
    custom_window_start_minutes: int = Field(default=0, ge=0, le=120)
    custom_window_end_minutes: int = Field(default=30, ge=0, le=240)
    day_of_week: Optional[int] = Field(None, ge=1, le=7, description="Day of week (1=Mon, 7=Sun)")
    time_slot_id: Optional[int] = Field(None, description="Time slot ID to create a schedule entry")


class CourseUpdate(BaseModel):
    """PUT /api/v1/courses/{id} — update an existing course."""

    course_name: Optional[str] = Field(None, min_length=1, max_length=255)
    course_code: Optional[str] = Field(None, max_length=50)
    department_id: Optional[uuid.UUID] = None
    room_id: Optional[uuid.UUID] = None
    teacher_id: Optional[int] = Field(None, description="Teacher ID from teachers table to reassign instructor")
    attendance_mode: Optional[Literal["preset", "flexible", "custom"]] = None
    custom_window_start_minutes: Optional[int] = Field(None, ge=0, le=120)
    custom_window_end_minutes: Optional[int] = Field(None, ge=0, le=240)
    day_of_week: Optional[int] = Field(None, ge=1, le=7)
    time_slot_id: Optional[int] = Field(None)


class CourseOut(BaseModel):
    """Read response for a single course."""

    id: uuid.UUID
    course_name: str
    course_code: Optional[str] = None
    teacher_id: int | None = None
    department_id: uuid.UUID | None = None
    room_id: uuid.UUID | None = None
    attendance_mode: Literal["preset", "flexible", "custom"] = "preset"
    custom_window_start_minutes: int = 0
    custom_window_end_minutes: int = 30
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None
    teacher_name: Optional[str] = None
    department_name: Optional[str] = None
    room_name: Optional[str] = None
    enrolled_count: int = 0
    day_of_week: Optional[int] = None
    time_slot_id: Optional[int] = None
    time_slot_name: Optional[str] = None

    model_config = {"from_attributes": True}


class CourseList(BaseModel):
    """Paginated wrapper for GET /api/v1/courses."""

    total: int
    items: list[CourseOut]


class CourseStudentDetail(BaseModel):
    """Single student with face status in a course."""

    student_id: int
    student_code: Optional[str] = None
    name: Optional[str] = None
    user_id: Optional[str] = None
    pin: Optional[str] = None
    has_face: bool
    embedding_count: int
    enrolled_at: datetime

    model_config = {"from_attributes": True}


class CourseStudentListResponse(BaseModel):
    """GET /api/v1/courses/{id}/students — student list with face status."""

    course_id: uuid.UUID
    total: int
    students: list[CourseStudentDetail]


class BatchEnrollRequest(BaseModel):
    """POST /api/v1/courses/{id}/students/batch — enroll multiple students."""

    student_ids: list[int] = Field(min_length=1, max_length=100)


class BatchEnrollResult(BaseModel):
    """Result for each student in batch enrollment."""

    student_id: int
    success: bool
    message: str


class BatchEnrollResponse(BaseModel):
    """Batch enrollment response."""

    course_id: uuid.UUID
    total_requested: int
    total_enrolled: int
    total_already_enrolled: int
    results: list[BatchEnrollResult]


class AvailableStudentDetail(BaseModel):
    """Student available for enrollment with face status."""

    id: int
    user_id: Optional[str] = None
    student_code: Optional[str] = None
    pin: Optional[str] = None
    full_name: Optional[str] = None
    has_face: bool
    embedding_count: int

    model_config = {"from_attributes": True}


class AvailableStudentListResponse(BaseModel):
    """GET /api/v1/courses/{id}/available-students — students not enrolled in course."""

    course_id: uuid.UUID
    total: int
    students: list[AvailableStudentDetail]

File: lib/entities/course.dart
enum AttendanceMode {
  preset,
  flexible,
  custom;

  String get value {
    switch (this) {
      case AttendanceMode.preset:
        return 'preset';
      case AttendanceMode.flexible:
        return 'flexible';
      case AttendanceMode.custom:
        return 'custom';
    }
  }

  String get label {
    switch (this) {
      case AttendanceMode.preset:
        return 'Tự động mở/đóng theo thời gian tiết học';
      case AttendanceMode.flexible:
        return 'Giáo viên đóng mở thủ công khi tiết đang diễn ra';
      case AttendanceMode.custom:
        return 'Mở/đóng theo khoảng thời gian tự thiết lập (trong tiết học)';
    }
  }

  String get shortLabel {
    switch (this) {
      case AttendanceMode.preset:
        return 'Cố định';
      case AttendanceMode.flexible:
        return 'Linh hoạt';
      case AttendanceMode.custom:
        return 'Tự thiết lập';
    }
  }

  static AttendanceMode fromString(String? value) {
    switch (value) {
      case 'flexible':
        return AttendanceMode.flexible;
      case 'custom':
        return AttendanceMode.custom;
      default:
        return AttendanceMode.preset;
    }
  }
}

class Course {
  final String id;
  final String courseName;
  final String? courseCode;
  final int? teacherId;
  final String? teacherName;
  final String? departmentId;
  final String? departmentName;
  final String? roomId;
  final String? roomName;
  final AttendanceMode attendanceMode;
  final int customWindowStartMinutes;
  final int customWindowEndMinutes;
  final int enrolledCount;
  final int? dayOfWeek;
  final int? timeSlotId;
  final String? timeSlotName;
  final int? totalSessions;
  final int? credits;
  final bool isCourseActiveNow;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Course({
    required this.id,
    required this.courseName,
    this.courseCode,
    this.teacherId,
    this.teacherName,
    this.departmentId,
    this.departmentName,
    this.roomId,
    this.roomName,
    this.attendanceMode = AttendanceMode.preset,
    this.customWindowStartMinutes = 0,
    this.customWindowEndMinutes = 30,
    this.enrolledCount = 0,
    this.dayOfWeek,
    this.timeSlotId,
    this.timeSlotName,
    this.totalSessions,
    this.credits,
    this.isCourseActiveNow = true,
    required this.createdAt,
    this.updatedAt,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'] as String,
      courseName: json['course_name'] as String,
      courseCode: json['course_code'] as String?,
      teacherId: json['teacher_id'] as int?,
      teacherName: json['teacher_name'] as String?,
      departmentId: json['department_id'] as String?,
      departmentName: json['department_name'] as String?,
      roomId: json['room_id'] as String?,
      roomName: json['room_name'] as String?,
      attendanceMode: AttendanceMode.fromString(
          json['attendance_mode'] as String? ?? 'preset'),
      customWindowStartMinutes:
          json['custom_window_start_minutes'] as int? ?? 0,
      customWindowEndMinutes:
          json['custom_window_end_minutes'] as int? ?? 30,
      enrolledCount: json['enrolled_count'] as int? ?? 0,
      dayOfWeek: json['day_of_week'] as int?,
      timeSlotId: json['time_slot_id'] as int?,
      timeSlotName: json['time_slot_name'] as String?,
      totalSessions: json['total_sessions'] as int?,
      credits: json['credits'] as int?,
      isCourseActiveNow: json['is_course_active_now'] ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'course_name': courseName,
      'course_code': courseCode,
      'teacher_id': teacherId,
      'department_id': departmentId,
      'room_id': roomId,
      'attendance_mode': attendanceMode.value,
      'custom_window_start_minutes': customWindowStartMinutes,
      'custom_window_end_minutes': customWindowEndMinutes,
      'day_of_week': dayOfWeek,
      'time_slot_id': timeSlotId,
      if (totalSessions != null) 'total_sessions': totalSessions,
      if (credits != null) 'credits': credits,
    };
  }

  Course copyWith({
    String? id,
    String? courseName,
    String? courseCode,
    int? teacherId,
    String? teacherName,
    String? departmentId,
    String? departmentName,
    String? roomId,
    String? roomName,
    AttendanceMode? attendanceMode,
    int? customWindowStartMinutes,
    int? customWindowEndMinutes,
    int? enrolledCount,
    int? dayOfWeek,
    int? timeSlotId,
    String? timeSlotName,
    int? totalSessions,
    int? credits,
    bool? isCourseActiveNow,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Course(
      id: id ?? this.id,
      courseName: courseName ?? this.courseName,
      courseCode: courseCode ?? this.courseCode,
      teacherId: teacherId ?? this.teacherId,
      teacherName: teacherName ?? this.teacherName,
      departmentId: departmentId ?? this.departmentId,
      departmentName: departmentName ?? this.departmentName,
      roomId: roomId ?? this.roomId,
      roomName: roomName ?? this.roomName,
      attendanceMode: attendanceMode ?? this.attendanceMode,
      customWindowStartMinutes:
          customWindowStartMinutes ?? this.customWindowStartMinutes,
      customWindowEndMinutes:
          customWindowEndMinutes ?? this.customWindowEndMinutes,
      enrolledCount: enrolledCount ?? this.enrolledCount,
      dayOfWeek: dayOfWeek ?? this.dayOfWeek,
      timeSlotId: timeSlotId ?? this.timeSlotId,
      timeSlotName: timeSlotName ?? this.timeSlotName,
      totalSessions: totalSessions ?? this.totalSessions,
      credits: credits ?? this.credits,
      isCourseActiveNow: isCourseActiveNow ?? this.isCourseActiveNow,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() =>
      'Course(id: $id, name: $courseName, code: $courseCode)';
}
